src/parse.cpp

```diff
@@ -264,10 +267,17 @@ expression* parser::parse_primary_expression(precedence prec) {
   // (x) => {}    // Arrow function.
   // (x + y * z)  // Parenthesized expression.
   case token_type::left_paren: {
+    this->depth_++;
+    if (this->depth_ > this->limit_) {
+      QLJS_PARSER_DEPTH_LIMIT_EXCEEDED();
+    }
     source_code_span left_paren_span = this->peek().span();
     this->skip();
 
     if (this->peek().type == token_type::right_paren) {
+      if (this->depth_ > 0) {
```

Comment:
When would depth be 0? Sounds like a bug. If anything, we should *assert* that the depth is >0:
```suggestion
      QLJS_ASSERT(this->depth_ > 0);
```